### ISSUE TITLE
"global" fully-loaded support for the Viewer

### ISSUE THREAD

**Issue by iangilman** [openseadragon/openseadragon#1257]
2.3.0 brought the "fully-loaded" support to the individual TiledImages, but it would probably be good to have a rolled up version for the Viewer. It would be the same event and accessor as with TiledImage, but it would only be true when all of the TiledImages in the viewer are fully loaded.

**Comment by iangilman** [comment]
Also, I think it's a common use case to want to do something after the image is fully loaded, so maybe we should also add a `whenFullyLoaded` to both the Viewer and the TiledImage. It would be something like so:

```
function whenFullyLoaded(tiledImage, callback) {
  if (tiledImage.getFullyLoaded()) {
    setTimeout(callback, 1); // So both paths are asynchronous
  } else {
    tiledImage.addOnceHandler('fully-loaded-change', function() {
      callback();
    });
  }
}
```

**Comment by iangilman** [comment]
Here's an example of how to make a viewer-level "fully loaded":

```
function areAllFullyLoaded() {
  var tiledImage;
  var count = viewer.world.getItemCount();
  for (var i = 0; i < count; i++) {
    tiledImage = viewer.world.getItemAt(i);
    if (!tiledImage.getFullyLoaded()) {
      return false;
    }
  }
  return true;
}

var isFullyLoaded = false;

viewer.world.addHandler('add-item', function(event) {
  var tiledImage = event.item;
  tiledImage.addHandler('fully-loaded-change', function() {
    var newFullyLoaded = areAllFullyLoaded();
    if (newFullyLoaded !== isFullyLoaded) {
      isFullyLoaded = newFullyLoaded;
      // Raise event
    }
  });
}
```

Of course this would have to be translated into the Viewer object to make a patch.

**Comment by achu1998** [comment]
If nobody is working on this, I can take it up.

**Comment by iangilman** [comment]
@achu1998 That would be wonderful... go for it!

**Comment by Aiosa** [comment]
(FYI) Just note that `fully loaded` event is one thing, but the new OSD data pipeline also makes the data update asynchronous. E.g., fact that data is fully loaded does not mean they are yet ready to render, if someone calls invalidation event there is also a promise (returned by invalidation call) available that resolves once the data is fully updated. Might/might not be relevant.

**Comment by iangilman** [comment]
@Aiosa Good point! Does that promise wait until everything is loaded and updated? 

I guess I'm thinking people can use that instead of this "fully loaded" if they are adding extra pipeline steps, and otherwise they can use this "fully loaded". Unless you can think of a good way to integrate them.

I think the general intention is to have an event that fires when all of the hi-rez tiles for a view are being displayed.

**Comment by achu1998** [comment]
@iangilman @Aiosa So we should consider both "fully loaded" as well as data pipeline update status (invalidation call) for the final fully loaded?

**Comment by iangilman** [comment]
@achu1998 I'm suggest we don't include the data pipeline update status in the new "fully loaded", but I'm open to further thoughts from @Aiosa.

**Comment by Aiosa** [comment]
It depends on what you want to achieve. I would say it's worth considering having an event for the finish of update event, not sure if it should be the same event. Also depends on whether it's desirable to fire it when the currently visible tiles are processed, or the whole cache data updated...

**Comment by iangilman** [comment]
@Aiosa Well, if what I wanted was to be notified when the current view was fully loaded and processed and drawn to the screen, how would you propose implementing that?

**Comment by Aiosa** [comment]
I would say if you don't fire invalidation of tiles, the fully-loaded event works as-is, since the async preparation is covered by the async tile caching nature. Not that easy with invalidation of existing data in the cache. But If I remember correctly the promise of the invalidation should resolves once all items were processed (or aborted), so if you don't have access to the promise object, it asks for an event fired before the resolution. Depending on logics of the invalidation (which was not decided or implemented yet), one might want to react on update of the visible part of the data, or all data in the cache. I think right now it resolves after all data has been updated  so there is no way of telling when the viewport data was updated. But I am now on train and did not look into the code so it also might be that the promise resolves after viewpoint data was updated and the rest happens -sometime after-.

**Comment by iangilman** [comment]
@Aiosa Cool, thank you.

@achu1998 Let's just stick with the "fully loaded" and not worry about the invalidation pipeline for now.

**Comment by Aiosa** [comment]
I just looked in the code, the invalidation is reused also for newly arriving tiles (to fire the pipeline on them too), so it would need some implementation to support this properly. I would also say that for now the invalidation can be ignored, if someone needs that they will surely raise an issue.

**Comment by achu1998** [comment]
@iangilman @Aiosa Thanks for that. I will proceed with "fully loaded" for now.

**Comment by iangilman** [comment]
@achu1998 Excellent. I look forward to seeing it in action!

**Comment by achu1998** [comment]
@iangilman I have added PR for this issue #2707
